### ISSUE TITLE
Make it easier for user to execute a system of capsules

### ISSUE THREAD

**Issue by dwtj** [hridesh/panini#89]
Currently, the standard way for the user to launch a system of capsules is by executing some `*$Thread` class as a Java application (either via the CLI or Eclipse). This generated class is tucked away alongside generated code, and ideally we want to keep all generated code away from the user's consideration.

We should have simpler way to launch a user's capsules. I am thinking that we could make a capsule start-able via some simple common static function call using reflection. Can we write such a function so that the user can do something like this:

```{.java}
package com.foo;

import org.paninij.lang.Capsule;

@Capsule
public class FooTemplate
{
    @Local Bar bar;
    Object field = new Object();

    public void run() {
        bar.doWork();
    }

    public static void main(String[] args) {
        Capsule.main("com.foo.Foo", args);
    }
}
```

This would need to work for both active and passive capsules.

I believe that having this would eliminate the need for manual main generation. We would basically let the user start their own capsule systems wherever they please.

**Comment by dwtj** [comment]
My initial hope is that we can implement such a `Capsule.main` by using reflection to take capsule's name, lookup a class, and create an instance of it.

We can also have a variant of `Capsule.main()` where the user additionally specifies a particular execution profile (rather than just using a default execution profile).

**Comment by dwtj** [comment]
I was hoping that we could add a static method to `Capsule`, but it looks like we can't add static methods to an annotation.

The existing place that seems similar in spirit which could support a static method is `Panini$System`, but that is *not* meant to be user facing (hence the `Panini$` prefix). How about making a new class: `org.paninij.lang.System`?

**Comment by dwtj** [comment]
A prototype of the above described functionality has been added as of commit 73be6b29cfb5b9df62b5eab193a4868a6681651d.

In order to get this functionality prototyped, all capsules have been made to "deserve" to have a `main()` method. This means that code-gen always adds a `main()` method. This has the advantage that it lets us sidestep the issue of how we should correctly and modularly implement this decision. Now it is basically up to the user to start capsule systems from whichever capsule he/she wants.

I think that this quick change to make all capsules "deserve" a main leaves some important questions for us, that I don't know how to answer right away:

- Was it correct to make *all* capsules deserve a main?

- It is clear that the user could start a "useless" capsule system (e.g. one which contains no active nodes), but what happens in these cases? Does our system behave gracefully? Is there a root node from which the user can start a capsule system which behaves (somehow) incorrectly?

- Based on how a capsule system is started in `CapsuleSystem.start()`, does a capsule system gracefully terminate when it should?

If we do indeed decide to make all capsules deserve a main, then there is refactoring to be done to simplify code-gen to remove the concept of "deserving a main". Even if we get PR #90 in its current form merged, I think that we should still keep this issue open until we clean up the code a bit more and have some better answers to these questions.

**Comment by dwtj** [comment]
See Issue #8 for more on this topic.

**Comment by dwtj** [comment]
After discussing this topic, we decided that we could sidestep some of these issues by using an `@Root` annotation to let the user decide whether a capsule "deserves a main". This functionality has been implemented and merged as part of PR #90.

Having resolved this issue, better garbage collection and more graceful capsule system shutdown should be easier to reason about in the future.
